File: 938.Range-Sum-of-BST.py
# https://leetcode.com/problems/range-sum-of-bst/description/
#
# algorithms
# Medium (84.21%)
# Total Accepted:    4.8k
# Total Submissions: 5.7k


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rangeSumBST(self, root, L, R):
        """
        :type root: TreeNode
        :type L: int
        :type R: int
        :rtype: int
        """
        res = [0]

        def resursive(node):
            if not node:
                return

            resursive(node.left)
            if L <= node.val <= R:
                res[0] += node.val

            resursive(node.right)

        resursive(root)

        return res[0]
        

File: 938.Range-Sum-of-BST.java
// https://leetcode.com/problems/range-sum-of-bst/
//
// algorithms
// Easy (78.31%)
// Total Accepted: 112,848
// Total Submissions: 144,098


/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    private static int res;

    public int rangeSumBST(TreeNode root, int L, int R) {
        res = 0;

        recursive(root, L, R);

        return res;
    }

    public void recursive(TreeNode root, int L, int R) {
        if (root == null) {
            return;
        }

        recursive(root.left, L, R);
        if (root.val >= L && root.val <= R) {
            res += root.val;
        } else if (root.val > R) {
            return;
        }
        recursive(root.right, L, R);
    }
}